src/main.cpp

```diff
@@ -90,87 +91,197 @@ namespace
+					// modify metadata in a guarded namespace.
+					return ERROR(CAT_INSUFFICIENT_PRIVILEGE_LEVEL, "User is not allowed to modify metadata");
+				}
+			}
+		}
+		catch (const json::exception&) {
+			// clang-format off
             log_re::error({{"log_message", "Unexpected JSON access or type error."},
                            {"rule_engine_plugin", "metadata_guard"}});
             // clang-format on
-        }
-        catch (const std::exception& e) {
-            log_re::error({{"log_message", e.what()}, {"rule_engine_plugin", "metadata_guard"}});
-        }
-
-        return CODE(RULE_ENGINE_CONTINUE);
-    }
+		}
+		catch (const std::exception& e) {
+			log_re::error({{"log_message", e.what()}, {"rule_engine_plugin", "metadata_guard"}});
+		}
+		return CODE(RULE_ENGINE_CONTINUE);
+	}
+
+	auto pep_atomic_metadata_handler(std::list<boost::any>& _rule_arguments, irods::callback _effect_handler)
+		-> irods::error
+	{
+		auto is_input_valid = [](const bytesBuf_t* _input) -> std::tuple<bool, std::string> {
+			if (!_input) {
+				return {false, "Missing JSON input"};
+			}
+
+			if (_input->len <= 0) {
+				return {false, "Length of buffer must be greater than zero"};
+			}
+
+			if (!_input->buf) {
+				return {false, "Missing input buffer"};
+			}
+
+			return {true, ""};
+		};
+
+		try {
+			auto* input_bb = boost::any_cast<bytesBuf_t*>(*std::next(std::begin(_rule_arguments), 2));
+			auto& rei = get_rei(_effect_handler);
+			const auto config = load_plugin_config(rei);
+
+			if (const auto [valid, msg] = is_input_valid(input_bb); !valid) {
+				log_re::error(msg);
+				return ERROR(INPUT_ARG_NOT_WELL_FORMED_ERR, msg);
+			}
+
+			const auto input_bb_casted = static_cast<const char*>(input_bb->buf);
+			json input = json::parse(input_bb_casted, input_bb_casted + input_bb->len);
+
+			const auto& operations = input.at("operations");
+
+			for (auto&& op : operations) {
+				const auto& err_code = metadata_guard_handling(op.at("attribute").get_ref<const std::string&>(), rei);
+				if (!err_code.ok()) {
+					return err_code;
+				}
+			}
+
+			return CODE(RULE_ENGINE_CONTINUE);
+		}
+		catch (const json::parse_error& e) {
+			log_re::error({{"log_message", "Failed to parse input into JSON"}, {"error_message", e.what()}});
+		}
+		catch (const std::exception& e) {
+			log_re::error({{"log_message", e.what()}, {"rule_engine_plugin", "metadata_guard"}});
+		}
+		return CODE(RULE_ENGINE_CONTINUE);
+	}
+
+	auto exec_rule(irods::default_re_ctx&,
+	               const std::string& _rule_name,
+	               std::list<boost::any>& _rule_arguments,
+	               irods::callback _effect_handler) -> irods::error
+	{
+		static const std::unordered_map<std::string,
+		                                std::function<irods::error(std::list<boost::any>&, irods::callback)>>
+			lookup_table{{"pep_api_mod_avu_metadata_pre", pep_modavu_handler},
+		                 {"pep_api_atomic_apply_metadata_operations_pre", pep_atomic_metadata_handler}};
```

Comment:
The handlers could use the full PEP names like so:
- `handle_pep_api_mod_avu_metadata_pre`
- `handle_pep_atomic_apply_metadata_operations_pre`

My advice is to always place common parts of the name at the front. That helps with sorting, grouping, and discovery _(for new devs)_.

Or perhaps you wrap the handlers in a well-named namespace (e.g. `pep_handler`).